### ISSUE TITLE
Socket Exception

### ISSUE THREAD

**Issue by oisgrafika** [localsend/localsend#125]
**
![image](https://user-images.githubusercontent.com/123151213/216894639-5c8136ee-c6d6-49e0-89a3-9b9a29bdf187.png)
**
 i get this error in windows

**Comment by itaowei** [comment]
![image](https://user-images.githubusercontent.com/19475586/220850744-f365df0b-5b88-4ea6-9ee8-3e3cf5cf9357.png)
Me too. Same error message...

**Comment by Tienisto** [comment]
(1) Try to start this app as admin
(2) Check in task manager if there is already another instance
(3) Use the latest version (`v1.7.0`) and check if it has been fixed

**Comment by itaowei** [comment]
> (1) Try to start this app as admin (2) Check in task manager if there is already another instance (3) Use the latest version (`v1.7.0`) and check if this has been fixed

I am sure that I followed all of your steps but it still goes wrong.

**Comment by TheStar4Fate** [comment]
> (1) Try to start this app as admin (2) Check in task manager if there is already another instance (3) Use the latest version (`v1.7.0`) and check if this has been fixed

Start with administrator privileges and still report an error

**Comment by hb-0** [comment]
+1

**Comment by xuyisoso** [comment]
> ![image](https://user-images.githubusercontent.com/19475586/220850744-f365df0b-5b88-4ea6-9ee8-3e3cf5cf9357.png) Me too. Same error message...

me too
版本	Windows 10 专业版
版本号	21H2
安装日期	‎2020/‎7/‎24
操作系统内部版本	19044.2075
体验	Windows Feature Experience Pack 120.2212.4180.0

**Comment by weijen6** [comment]
Me too, but I changed the port to 52319 and successfully connected

**Comment by grypes** [comment]
Same with @weijen6 , I changed the default port to another random port then it's working.

**Comment by Tienisto** [comment]
53317 seem to be a bad choice as there is a small possibility that it is used as a "ephemeral port" (temporary port).

I guess a proper fix for this is to use a lower number port and use 53317 as fallback.

**Comment by grypes** [comment]
Looks like this default port 53317 is reserved by Windows and other potential applications.
Below is the port reserved list of my runtime.  (Windows 11 Pro 1000.22640.1000.0)

So any port in the below range will run into the same issue. In my case, I changed to 59998 and it works well.

--------------------------

>netsh interface ipv4 show excludedportrange protocol=tcp

Protocol tcp Port Exclusion Ranges

Start Port    End Port
----------    --------
      5357        5357
     49675       49774
     49875       49974
     50000       50059     *
     50060       50159
     52683       52782
     52814       52913
     53287       53386
     59865       59964
     62040       62139
     62140       62239
     62240       62339
     62340       62439
     62440       62539
     62540       62639
     62640       62739

**Comment by zhaiyusci** [comment]
I have the same problem, but I find that the port 53317 is like blocked by localsend itself...
Using `netstat -ano|findstr 53317` in `cmd` and I can find the PID of the process blocking the port, and right after that I killed it, localsend quitted.

**Comment by alisonjoe** [comment]
Phone2Mac:
DioError [unknown]: null
Error: SocketException: No route to host (OS Error: No route to host, errno = 113), address = 172.16.4.33, port = 39840

Mac2Phone:
DioError [connection timeout]: The request connection took longer than 720:00:00.000000. It was aborted.
Error: SocketException: Operation timed out (OS Error: Operation timed out, errno = 60), address = 172.16.221.61, port = 56128

**Comment by tintad** [comment]
I had the same problem and fixed it by creating an inbound rule in Firewall settings for this app. It's working perfectly now
![LocalSend](https://github.com/localsend/localsend/assets/94805705/9ac8abda-2ddb-45b8-853f-2c8529c1b1d1)

**Comment by chenyi8920** [comment]
> Me too, but I changed the port to 52319 and successfully connected

much appreciated!

**Comment by zeyios** [comment]
Changing to another port number like 52199 on both devices works for me.

**Comment by zhangj1164** [comment]
> Looks like this default port 53317 is reserved by Windows and other potential applications. Below is the port reserved list of my runtime. (Windows 11 Pro 1000.22640.1000.0)
> 
> So any port in the below range will run into the same issue. In my case, I changed to 59998 and it works well.
> 
> > netsh interface ipv4 show excludedportrange protocol=tcp
> 
> Protocol tcp Port Exclusion Ranges
> 
> Start Port End Port ---------- -------- 5357 5357 49675 49774 49875 49974 50000 50059 * 50060 50159 52683 52782 52814 52913 53287 53386 59865 59964 62040 62139 62140 62239 62240 62339 62340 62439 62440 62539 62540 62639 62640 62739

3Q，this is correct. I checked my TCP port Exclusion Ranges and it is different.

**Comment by BIgFish420** [comment]
> (1) Try to start this app as admin
> (2) Check in task manager if there is already another instance
> (3) Use the latest version (`v1.7.0`) and check if this has been fixed

Very good keep up the good work

**Comment by BTBob53** [comment]
52319 worked for me

**Comment by zznyjidi** [comment]
I had the same issue and I tried
```
net stop winnat
```
It works for me. 
(Found at https://superuser.com/a/1610009)

**Comment by shzang** [comment]
> I had the same issue and I tried
> 
> ```
> net stop winnat
> ```
> 
> It works for me. (Found at https://superuser.com/a/1610009)

Thanks bro, after running this code, I can use port 53317 to transfer files normally.

**Comment by UnknownTI** [comment]
> Me too, but I changed the port to 52319 and successfully connected

Me too. Thanks!

**Comment by angelicachao** [comment]
Just faced this issue as well, restarting Internet Connection Sharing from services fixed it immediately, didn't even need to restart localsend, was able to start the server from localsend's settings right away.

**Comment by flt6** [comment]
> Looks like this default port 53317 is reserved by Windows and other potential applications. Below is the port reserved list of my runtime. (Windows 11 Pro 1000.22640.1000.0)
> 
> So any port in the below range will run into the same issue. In my case, I changed to 59998 and it works well.
> 
> > netsh interface ipv4 show excludedportrange protocol=tcp
> 
> Protocol tcp Port Exclusion Ranges
> 
> Start Port End Port ---------- -------- 5357 5357 49675 49774 49875 49974 50000 50059 * 50060 50159 52683 52782 52814 52913 53287 53386 59865 59964 62040 62139 62140 62239 62240 62339 62340 62439 62440 62539 62540 62639 62640 62739

Same case for Win 10 22H2. `53291       53390` is reserved on my computer.

**Comment by rbjnazz** [comment]
If you're using windows:
Search `Network status` then find `Network reset` click then restart your pc.

WARNING: This would erase saved wifi passwords and other network settings.

**Comment by AneryCoft** [comment]
![image](https://github.com/user-attachments/assets/2c16133f-ef28-45f2-a471-13bc29546444)
I faced this issue too.
I changed port and it worked now
btw dont forget change port on Android

**Comment by AneryCoft** [comment]
![image](https://github.com/user-attachments/assets/2c16133f-ef28-45f2-a471-13bc29546444)
I faced this issue too.
I changed port and it worked now
btw dont forget change port on Android

**Comment by r0hitm** [comment]
Just encountered this. I just rebooted the PC, and it fixed it

**Comment by TheSameCat2** [comment]
Just adding a data point:
Just encountered this bug on Windows 11 Pro, 23H2. I do have Hyper-V installed on this machine (I notice in the thread that Hyper-V has been brought up.

I fixed it by changing both the desktop and my iOS device to port 52319 as mentioned in this thread.

**Comment by SiWuXuan** [comment]
change port and it will work as usual

**Comment by ghmole** [comment]
win 19 22H2 can't run 
 以一种访问权限不允许的方式做了一个访问套接字的尝试

**Comment by Tienisto** [comment]
The primary reason for this to happen more often than expected is because the high port numbers are used as "ephemeral ports". 

I think https://github.com/localsend/protocol/issues/12 will resolve this issue.

**Comment by adam-ah** [comment]
@Tienisto a single fallback does not solve it, manually changed to 53318, or 53319, or 53320, or 53321, or 53330, 53340, 53350 but exact same error.

Nothing is occupying these ports on my system, these are *reserved for Windows*:
```
> netsh interface ipv4 show excludedportrange protocol=tcp

Protocol tcp Port Exclusion Ranges

Start Port    End Port
----------    --------
     49693       49792
     49793       49892
     49893       49992
     50000       50059     *
     50060       50159
     50160       50259
     51113       51212
     51585       51684
     51685       51784
     51785       51884
     51885       51984
     51985       52084
     52085       52184
     53087       53186
     53217       53316
     53317       53416
     53417       53516
```
Jumping more helps, but LocalSend either needs to parse this exclusion table, or simply keep jumping until a port is found.

**Comment by baviereteam** [comment]
Note that you can add an exception for port 53317 this way (in an elevated Powershell or CMD):
```
net stop winnat
netsh int ipv4 add excludedportrange protocol=tcp startport=53317 numberofports=1
net start winnat
```

When running `netsh interface ipv4 show excludedportrange protocol=tcp` again, you should see that port 53317 appears on its own line and with a star (that, I guess, means that is an exception you added yourself) :
```
PS C:\Users\xxx> netsh interface ipv4 show excludedportrange protocol=tcp

Protocole tcp Plages d'exclusion de ports

Port de début    Port de fin
-------------    -----------
[...]
     53317       53317     *
[...]

* - Exclusions de ports administrés.
```

**Comment by dhrm1k** [comment]
@baviereteam your solution did work? can you explain what did it do?

**Comment by Stmsmj** [comment]
I went to network connections in control panel and disabled and enabled my connections and problem got solved

**Comment by griessel-jc** [comment]
> I went to network connections in control panel and disabled and enabled my connections and problem got solved

This solved it for me, thank you.

**Comment by dhrm1k** [comment]
[baviereteam](https://github.com/baviereteam) your solution works but i also have wsl2 installed on my machine. After using your solution, wsl2 couldn't connect to the internet.

**Comment by hamzamg** [comment]
> Note that you can add an exception for port 53317 this way (in an elevated Powershell or CMD):
> 
> ```
> net stop winnat
> netsh int ipv4 add excludedportrange protocol=tcp startport=53317 numberofports=1
> net start winnat
> ```
> 
> When running `netsh interface ipv4 show excludedportrange protocol=tcp` again, you should see that port 53317 appears on its own line and with a star (that, I guess, means that is an exception you added yourself) :
> 
> ```
> PS C:\Users\xxx> netsh interface ipv4 show excludedportrange protocol=tcp
> 
> Protocole tcp Plages d'exclusion de ports
> 
> Port de début    Port de fin
> -------------    -----------
> [...]
>      53317       53317     *
> [...]
> 
> * - Exclusions de ports administrés.
> ```

is WORK

**Comment by InternetNinjo** [comment]
> Note that you can add an exception for port 53317 this way (in an elevated Powershell or CMD):
> 
> ```
> net stop winnat
> netsh int ipv4 add excludedportrange protocol=tcp startport=53317 numberofports=1
> net start winnat
> ```
> 
> When running `netsh interface ipv4 show excludedportrange protocol=tcp` again, you should see that port 53317 appears on its own line and with a star (that, I guess, means that is an exception you added yourself) :
> 
> ```
> PS C:\Users\xxx> netsh interface ipv4 show excludedportrange protocol=tcp
> 
> Protocole tcp Plages d'exclusion de ports
> 
> Port de début    Port de fin
> -------------    -----------
> [...]
>      53317       53317     *
> [...]
> 
> * - Exclusions de ports administrés.
> ```

isn't winnat only in vm's, it doesn't exist on windows 11 23h2 :/

**Comment by InternetNinjo** [comment]
> Note that you can add an exception for port 53317 this way (in an elevated Powershell or CMD):
> 
> ```
> net stop winnat
> netsh int ipv4 add excludedportrange protocol=tcp startport=53317 numberofports=1
> net start winnat
> ```
> 
> When running `netsh interface ipv4 show excludedportrange protocol=tcp` again, you should see that port 53317 appears on its own line and with a star (that, I guess, means that is an exception you added yourself) :
> 
> ```
> PS C:\Users\xxx> netsh interface ipv4 show excludedportrange protocol=tcp
> 
> Protocole tcp Plages d'exclusion de ports
> 
> Port de début    Port de fin
> -------------    -----------
> [...]
>      53317       53317     *
> [...]
> 
> * - Exclusions de ports administrés.
> ```

WORKS BTW

**Comment by cooolbrains** [comment]
> Me too, but I changed the port to 52319 and successfully connected

This works 100%

**Comment by InternetNinjo** [comment]
> [@baviereteam](https://github.com/baviereteam) your solution works but i also have wsl2 installed on my machine. After using your solution, wsl2 couldn't connect to the internet.

use another available port ig? different for wsl2 and different for the method

**Comment by gongjimin** [comment]
Thanks 

> > Me too, but I changed the port to 52319 and successfully connected
> 
> This works 100%

Thanks a lot.

**Comment by JdotCarver** [comment]
> Note that you can add an exception for port 53317 this way (in an elevated Powershell or CMD):
> 
> ```
> net stop winnat
> netsh int ipv4 add excludedportrange protocol=tcp startport=53317 numberofports=1
> net start winnat
> ```
> 
> When running `netsh interface ipv4 show excludedportrange protocol=tcp` again, you should see that port 53317 appears on its own line and with a star (that, I guess, means that is an exception you added yourself) :
> 
> ```
> PS C:\Users\xxx> netsh interface ipv4 show excludedportrange protocol=tcp
> 
> Protocole tcp Plages d'exclusion de ports
> 
> Port de début    Port de fin
> -------------    -----------
> [...]
>      53317       53317     *
> [...]
> 
> * - Exclusions de ports administrés.
> ```

Yet another quote to say : This works !
Would love LocalSend to say a little :  
"Hey user! I see that you are on Windows. You should run these 3 commands in order to authorize the port."
Or even better, make it automatic on first setup. 😃

**Comment by nhblock** [comment]
> Note that you can add an exception for port 53317 this way (in an elevated Powershell or CMD):
> 
> ```
> net stop winnat
> netsh int ipv4 add excludedportrange protocol=tcp startport=53317 numberofports=1
> net start winnat
> ```
> 
> When running `netsh interface ipv4 show excludedportrange protocol=tcp` again, you should see that port 53317 appears on its own line and with a star (that, I guess, means that is an exception you added yourself) :
> 
> ```
> PS C:\Users\xxx> netsh interface ipv4 show excludedportrange protocol=tcp
> 
> Protocole tcp Plages d'exclusion de ports
> 
> Port de début    Port de fin
> -------------    -----------
> [...]
>      53317       53317     *
> [...]
> 
> * - Exclusions de ports administrés.
> ```

Confirming this method works to resolve the error. Changing the port did not.
Windows 11 Pro ARM on Surface Snapdragon

**Comment by user1823** [comment]
I was getting this error:

SocketException: Failed to create server socket (OS Error: Only one usage of each socket address (protocol/network address/port) is normally permitted. , errno = 10048), address = 0.0.0.0, port = 53317

Changing the port to 52319 resolved the issue.

But, this means that I need to change the port on every device, which may not be very practical, especially if you want to share files with other people too. 

So, I tried this:
```
net stop winnat
netsh int ipv4 add excludedportrange protocol=tcp startport=53317 numberofports=1
net start winnat
```

After running this code and reverting the port to the default (53317), I was getting the same error.

Then, I ran `wsl --shutdown` in cmd and restarted the LocalSend server. This solved the issue. Now, both LocalSend and WSL2 are working fine.

Update:
This is not very reliable. Even though 53317 is in the list of excludedportrange, I still have to shut down wsl occasionally by using `wsl --shutdown` to free up the port.

**Comment by xqb-dpx** [comment]
**These ports are especially great (range 49152 - 65535):**

- 50000
- 50505
- 55000
- 60000
- 65000

These ports usually have no conflict for TCP on any OS

**Comment by MattAirhart** [comment]
> Note that you can add an exception for port 53317 this way (in an elevated Powershell or CMD):
> 
> ```
> net stop winnat
> netsh int ipv4 add excludedportrange protocol=tcp startport=53317 numberofports=1
> net start winnat
> ```

+1 to this method! Worked just fine for Windows 10 x64 22H2.
Maybe it should be considered for the Troubleshooting section of the readme

**Comment by netbuffer** [comment]
> Note that you can add an exception for port 53317 this way (in an elevated Powershell or CMD):
> 
> ```
> net stop winnat
> netsh int ipv4 add excludedportrange protocol=tcp startport=53317 numberofports=1
> net start winnat
> ```
> 
> When running `netsh interface ipv4 show excludedportrange protocol=tcp` again, you should see that port 53317 appears on its own line and with a star (that, I guess, means that is an exception you added yourself) :
> 
> ```
> PS C:\Users\xxx> netsh interface ipv4 show excludedportrange protocol=tcp
> 
> Protocole tcp Plages d'exclusion de ports
> 
> Port de début    Port de fin
> -------------    -----------
> [...]
>      53317       53317     *
> [...]
> 
> * - Exclusions de ports administrés.
> ```

worked for me

---
版本	Windows 11 专业版
版本号	25H2
操作系统版本	26200.8246

LocalSend版本
1.17.0（58）

**Comment by adam-ah** [comment]
It’s both frustrating and disheartening to see users resorting to manually opening ports in Windows, rather than LocalSend implementing automatic port selection and fallback mechanisms. No one should have to create manual firewall exceptions simply because an application has chosen an inappropriate or unavailable port to bind to.

**Comment by mimodavila-ops** [comment]
> It’s both frustrating and disheartening to see users Duplicate of #3067 
resorting to manually opening ports in Windows, rather than LocalSend implementing automatic port selection and fallback mechanisms. No one should have to create manual firewall exceptions simply because an application has chosen an inappropriate or unavailable port to bind to.

> 

> > Note that you can add an exception for port 53317 this way (in an elevated Powershell or CMD):
> > ```
> > net stop winnat
> > netsh int ipv4 add excludedportrange protocol=tcp startport=53317 numberofports=1
> > net start winnat
> > ```
> 
> +1 to this method! Worked just fine for Windows 10 x64 22H2. Maybe it should be considered for the Troubleshooting section of the readme

**Comment by ghost** [comment]
> > It’s both frustrating and disheartening to see users Duplicate of [#3067](https://github.com/localsend/localsend/pull/3067)
> > resorting to manually opening ports in Windows, rather than LocalSend implementing automatic port selection and fallback mechanisms. No one should have to create manual firewall exceptions simply because an application has chosen an inappropriate or unavailable port to bind to.
> 
> > 
> 
> > > Note that you can add an exception for port 53317 this way (in an elevated Powershell or CMD):
> > > ```
> > > net stop winnat
> > > netsh int ipv4 add excludedportrange protocol=tcp startport=53317 numberofports=1
> > > net start winnat
> > > ```
> > 
> > 
> > +1 to this method! Worked just fine for Windows 10 x64 22H2. Maybe it should be considered for the Troubleshooting section of the readme

Actually it's addressed here
https://github.com/localsend/localsend#setup

**Comment by Xqni** [comment]
had the same issue with the forbidden access and port 53317, simply running `wsl --shutdown` helps. i dont really use localsend that often but shutting wsl for once in a while doesn't bother my workflows.

p.s: i believe this is only happening because wsl has something to do with port 53317 (not so sure tho)

**Comment by SynapseScribe** [comment]
On https://learn.microsoft.com/en-us/troubleshoot/windows-server/networking/service-overview-and-network-port-requirements
it is clearly specified that Hyper-v services reserve ports from 49152 - 65535, making localsend unusable on default 53317 port. The randomly reserved ports can be seen through pwsh, in my case:

PS> netsh int ipv4 show excludedportrange protocol=tcp

Protocol tcp Port Exclusion Ranges

Start Port    End Port
----------    --------
      5985        5985
     47001       47001
     49673       49772
     50000       50059     *
     50060       50159
     50164       50263
     50366       50465
     50466       50565
     50566       50665
     53286       53385
     53386       53485

* - Administered port exclusions.
